### PR TITLE
Issue #5491: Check for wrong file paths in .tests.info file

### LINKED ISSUE #5491: Simpletest semi-silently ignores wrong file paths in .tests.info file
As noted in https://github.com/backdrop/backdrop-issues/issues/5490, 

> it turned out that several tests actually never run:
> core/modules/system/tests/system.tests.info contains wrong file paths
> It's still not clear, why these tests ran in the past, but don't for several weeks. Eventually related: Issue https://github.com/backdrop/backdrop-issues/issues/5415: Gracefully handle empty or malformed .tests.info files.

In https://github.com/backdrop/backdrop-issues/issues/5415, we introduced code that checked for malformed .tests.info files and put errors into watchdog when they're detected. We can (should) add to this error-checking to verify that the file paths specified in the tests.info file are correct.

PR to follow.

### PR DESCRIPTION
**Pull request by bugfolder**
Fixes https://github.com/backdrop/backdrop-issues/issues/5491.

Rebase of https://github.com/backdrop/backdrop/pull/4249.

### CODE REVIEW

File: core/modules/simpletest/simpletest.module [hunk 1]
```diff
@@ -411,8 +424,17 @@ function simpletest_test_get_all() {
...
-          watchdog('simpletest', $error, array('%uri' => $file->uri), WATCHDOG_WARNING, l(t('See the Testing Framework documentation.'), 'https://docs.backdropcms.org/documentation/testing-framework'));
-          continue;
+          if (backdrop_is_cli()) {
+            // If running in CLI, then we're probably about to run some
+            // automated tests, so we need a hard fail.
+            print format_string($error, $variables) . "\n";
+            exit(1);
+          }
+          else {
+            backdrop_set_message(format_string($error, $variables), 'error');
```
Comment (klonos) [inline core/modules/simpletest/simpletest.module:441]: As I've elaborated in detail in my comment in the issue, this is not proper use of `t()` AFAIK. `$error` should remain a plain string (non-`t()`'ed), with the substitution placeholders, and then you should be doing this here:

```suggestion
            backdrop_set_message(t($error, $variables), 'error');
```

File: core/modules/simpletest/simpletest.module [hunk 2]
```diff
@@ -411,8 +424,17 @@ function simpletest_test_get_all() {
...
+            watchdog('simpletest', $error, $variables, WATCHDOG_WARNING, l(t('See the Testing Framework documentation.'), 'https://docs.backdropcms.org/documentation/testing-framework'));
```
Comment (klonos) [inline core/modules/simpletest/simpletest.module:442]: No need to do `t($error)` here 👍🏼

File: core/modules/simpletest/simpletest.module [hunk 3]
```diff
@@ -369,17 +369,19 @@ function simpletest_test_get_all() {
...
+          // informative error.
           foreach ($classes as $class => $info) {
-            // If the .tests.info file is malformed, it puts a slew of
-            // hard-to-diagnose errors into the watchdog log when the testing page
-            // is loaded. So we'll check for various malformalities and put a more
-            // informative error.
             if (!is_array($info) || empty($info)) {
-              $error = format_string('File %uri contains a definition for %class that is not an array of testing class fields.', array('%class' => $class));
+              $error = t('File @uri contains a definition for @class that is not an array of testing class fields.');
+              $variables += array('@class' => $class);
```
Comment (klonos) [inline core/modules/simpletest/simpletest.module:389]: If you want to use `t()` here, then you should be doing something like this:
```suggestion
              $variables += array('@class' => $class);
              $error = t('File @uri contains a definition for @class that is not an array of testing class fields.', $variables);
```

File: core/modules/simpletest/simpletest.module [hunk 4]
```diff
@@ -391,13 +393,24 @@ function simpletest_test_get_all() {
               }
               if (!empty($missing_keys)) {
                 $keys = implode(', ', $missing_keys);
-                $error = format_plural(count($missing_keys), 'File %uri contains a definition for %class that is missing field %keys.','File %uri contains a definition for %class that is missing fields %keys.', array('%class' => $class, '%keys' => $keys));
+                $error = format_plural(count($missing_keys), 'File @uri contains a definition for @class that is missing field @keys.', 'File @uri contains a definition for @class that is missing fields @keys.');
+                $variables += array('@class' => $class, '@keys' => $keys);
```
Comment (klonos) [inline core/modules/simpletest/simpletest.module:402]: ...same as here, but no need for `t()` inside `format_plural()`:
```suggestion
                $variables += array('@class' => $class, '@keys' => $keys);
                $error = format_plural(count($missing_keys), 'File @uri contains a definition for @class that is missing field @keys.', 'File @uri contains a definition for @class that is missing fields @keys.', $variables);
```

File: core/modules/simpletest/simpletest.module [hunk 5]
```diff
@@ -391,13 +393,24 @@ function simpletest_test_get_all() {
...
             }
+            // Verify that the test file is present in the same directory as
+            // specified in the .tests.info file.
             if (empty($error)) {
-            // If the format passed, build the rest of the groups information.
               $info['file path'] = dirname($file->uri);
+              $test_file_full = BACKDROP_ROOT . '/' . $info['file path'] . '/' . $info['file'];
+              if (!file_exists($test_file_full)) {
+                $error = t('File @uri specifies a non-existent file @test_file_full. Please verify that the test file exists in the location specified in @uri.');
+                $variables += array('@test_file_full' => $test_file_full);
```
Comment (klonos) [inline core/modules/simpletest/simpletest.module:413]:
```suggestion
                $variables += array('@test_file_full' => $test_file_full);
                $error = t('File @uri specifies a non-existent file @test_file_full. Please verify that the test file exists in the location specified in @uri.', $variables);
```

File: core/modules/simpletest/simpletest.module [hunk 6]
```diff
@@ -411,8 +424,17 @@ function simpletest_test_get_all() {
           }
         }
         if (!empty($error)) {
-          watchdog('simpletest', $error, array('%uri' => $file->uri), WATCHDOG_WARNING, l(t('See the Testing Framework documentation.'), 'https://docs.backdropcms.org/documentation/testing-framework'));
-          continue;
+          if (backdrop_is_cli()) {
+            // If running in CLI, then we're probably about to run some
+            // automated tests, so we need a hard fail.
+            print format_string($error, $variables) . "\n";
```
Comment (klonos) [inline core/modules/simpletest/simpletest.module:435]: `$error` should have already been `t()`'ed here:

```suggestion
            print $error . "\n";
```

File: core/modules/simpletest/simpletest.module [hunk 7]
```diff
@@ -369,17 +369,19 @@ function simpletest_test_get_all() {
       foreach ($files as $file) {
         $classes = backdrop_parse_info_file($file->uri, TRUE);
         $error = '';
+        $variables = array('@uri' => $file->uri);
         if (empty($classes)) {
-          $error = 'File %uri contains no testing classes.';
+          $error = t('File @uri contains no testing classes.');
```
Comment (klonos) [inline core/modules/simpletest/simpletest.module:379]:
```suggestion
          $error = t('File @uri contains no testing classes.', $variables);
```

File: core/modules/simpletest/simpletest.module (hunk 1 above)
Comment (klonos) [reply core/modules/simpletest/simpletest.module:441]: ...but if you already `t()` `$error` as in my suggestions above in this PR, then you can simply do this here:
```suggestion
            backdrop_set_message($error, 'error');
```

**Comment by quicksketch** [comment]
Merged into 1.x and 1.23.x. Thanks so much @bugfolder!

**Pull request merged**